be/src/io/s3_input_stream.cpp

```diff
@@ -41,19 +43,22 @@ StatusOr<int64_t> S3InputStream::read(void* out, int64_t count) {
         return 0;
     }
 
+    auto real_length = std::min<int64_t>(_offset + count, _size) - _offset;
+
     // https://www.rfc-editor.org/rfc/rfc9110.html#name-range
-    auto range = fmt::format("bytes={}-{}", _offset, std::min<int64_t>(_offset + count, _size) - 1);
+    auto range = fmt::format("bytes={}-{}", _offset, _offset + real_length - 1);
     Aws::S3::Model::GetObjectRequest request;
     request.SetBucket(_bucket);
     request.SetKey(_object);
     request.SetRange(std::move(range));
+    request.SetResponseStreamFactory([out, real_length]() {
+        return Aws::New<S3ZeroCopyIOStream>(AWS_ALLOCATE_TAG, reinterpret_cast<char*>(out), real_length);
+    });
 
     Aws::S3::Model::GetObjectOutcome outcome = _s3client->GetObject(request);
     if (outcome.IsSuccess()) {
-        Aws::IOStream& body = outcome.GetResult().GetBody();
-        body.read(static_cast<char*>(out), count);
-        _offset += body.gcount();
-        return body.gcount();
+        _offset += real_length;
```

Comment:
Why not reuse `outcome.GetResult().GetBody().gcount()`? Can we make sure it always equal to `real_length`?

Reply:
iostream's gcount is set after extract operation(like read), we didn't call read at all(we used to call body.read(xxx), and we can set the correct offset even if we had a wrong range #49692)